Lib/ast.py

```diff
@@ -613,6 +613,30 @@ def __exit__(self, exc_type, exc_value, traceback):
     def block(self):
         return self._Block(self)
 
+    class _NoDelimit:
+        def __enter__(self): pass
+        def __exit__(self, *args): pass
+
+    class _Delimit:
+        """A context manager for preparing the source for expressions. It adds
+        start of the delimiter  and enters, after exit it adds delimiter end."""
+        def __init__(self, unparser, delimiter):
+            self.unparser = unparser
+            self.delimiter = delimiter
+        def __enter__(self):
+            self.unparser.write(self.delimiter[0])
+        def __exit__(self, exc_type, exc_value, traceback):
+            self.unparser.write(self.delimiter[-1])
+
+    def delimit_if(self, condition, delimiter = "()"):
+        if condition:
+            return self._Delimit(self, delimiter)
+        else:
+            return self._NoDelimit()
+
+    def delimit(self, delimiter = "()"):
+        return self._Delimit(self, delimiter)
```

Comment:
IHMO a contextlib.contextmanager would be simpler. Something like:

```
@contextlib.contextmanager
def delim_if(self, condition, delimiter="()"):
    if condition:
        self.write(self.delimiter[0]) 
    try:
        yield
    finally:
        if condition:
            self.write(self.delimiter[-1])

# I'm not sure that this one works as expected :-)
def delim(self, delimiter):
    return self.delim_if(True, delimiter)
```

Reply:
We started with this, but it introduced a regression in the import time and I reverted it.

Reply:
https://github.com/python/cpython/pull/17376

Reply:
You wouldn't have such issue if ast.unparse would be a submodule....

Reply:
I dislike the submodule idea more:

https://github.com/python/cpython/pull/17377#discussion_r358010526

Reply:
This discussion is splitted between 3 comment threads in 2 PRs. I created https://bugs.python.org/issue39069 to summarize the discussion and discuss the idea there.